### PR TITLE
DM-5750: Bokeh server integration

### PR DESCRIPTION
**Pull request by afausti**
@jonathansick @frossie the integration of the dashboard app with the bokeh-server is done. Added  some hardcoded values to illustrate the time series plot.

### CODE REVIEW

**Comment by jonathansick** [comment]
I've just started testing it. It was quite easy to get it into dev mode thanks to your README, requirements.txt and startup scripts. 👍 

The Django REST api browser is also very cool. It's nice that one can add data from a web form in addition to HTTP commands sending HTTP commands from the command line. 👍 

Bokeh definitely works. I just notice that we'll have to tweak the figure sizing to make it responsive. If my browser window isn't super-wide then this happens:

napkin 16-05-04 1 44 35 pm

So the Bokeh UI is responding to the window size, but the plot itself isn't. Do you see a way to fix this off the top of your head?

**Comment by jonathansick** [comment]
The hover tool-tips for build info work great 👍

**Comment by afausti** [comment]
Hi @jonathansick you can try the the option 

plot = Figure(
        plot_height=400,
        plot_width=900,
        responsive=True,

in plots.py. However, I noticed a strange behaviour, when you have responsive=True and use the reset tool the plot size changes...

**Comment by jonathansick** [comment]
Just a thought that we should have a bootstrap script that takes a bunch of real validate_drp data and inserts it into the DB rather than needing to hard-code it for testing. No need to change this now though; probably we should switch to that mode by the next PR.

**Comment by afausti** [comment]
I agree, we need some test data to initialize the db and actually read the data from there, I think that is the next thing to do

**Comment by jonathansick** [comment]
Ah, setting `responsive=True` is awesome. We should file an upstream GitHub issue on the behaviour of 'reset' with a responsive plot.

File: squash/dashboard/templates/dashboard/metric.html [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        <div class="col-xs-10">
+            <!-- Using a fixed dataset for now, later this could be a pull down menu for selecting different datasets -->
+
+            <h4 class="pull-right">
+                <div class="dropdown">
+                    <button class="btn btn-default dropdown-toggle" type="button" data-toggle="dropdown">ci_hsc
+                        <span class="caret"></span></button>
+                    <ul class="dropdown-menu">
+                        <li><a href="#">ci_hsc</a></li>
+                        <li><a href="#">ci_cfht</a></li>
```
Comment (jonathansick) [inline squash/dashboard/templates/dashboard/metric.html]: I noticed this dropdown doesn't actually drop down to show the `ci_cfht` option.

**Comment by afausti** [comment]
OK will fix that, I guess the problem is the dropdown div inside the <h4/>

**Comment by jonathansick** [comment]
I think `viz/`needs an `__init__.py`?

`viz/plots.py` might be better renamed to `viz/defaults.py`. I'd re-name `time_series_defaults` to `init_time_series_plot` and `legend_defaults` to `init_legend` (make functions verbs).

File: squash/dashboard/views.py [hunk 2]
```diff
@@ -35,3 +42,24 @@ class MetricViewSet(DefaultsMixin, viewsets.ModelViewSet):
...
+class HomeView(TemplateView):
+    template_name = 'dashboard/index.html'
+
+
+class MetricDashboardView(DetailView):
+    template_name = 'dashboard/metric.html'
+    model = User
+
+    def get_context_data(self, **kwargs):
+        context = super(MetricDashboardView, self).get_context_data(**kwargs)
```
Comment (jonathansick) [inline squash/dashboard/views.py]: I think in Python 3 you can just do `super().get_context_data(**kwargs)`

Comment (timj) [inline squash/dashboard/views.py]: Yes, and this code can be all python 3 unless you are dependent on stack code.

**Comment by afausti** [comment]
yes, it is meant to be all python 3. thanks for the tips

**Comment by jonathansick** [comment]
Okay, I think this all looks good. I don't quite understand sessions yet. How is a user who isn't logged in identified?

**Comment by afausti** [comment]
The app still needs user authentication, for now only the user created during the db initialization is associated to the bokeh sessions, they are saved in the UserSession table

**Comment by jonathansick** [comment]
> The app still needs user authentication, for now only the user created during the db initialization is associated to the bokeh sessions, they are saved in the UserSession table

Ok. Will the Bokeh plotting depend on a user being logged in? Or can anonymous viewing also work? (not a complaint either way, just trying to understand how it will work)

**Comment by afausti** [comment]
Its per user, each user logged in the system will have its own bokeh session.

**Comment by jonathansick** [comment]
Okay, at some point we might have to identify anonymous user sessions if we make a strategic decision to allow anonymous browsing. At this stage I think it's all good. Consider it 'reviewed' 👍

**Comment by afausti** [comment]
@jonathansick thank you so much, will apply the suggestions and close the ticket!

**Pull request merged**